AGX Orin instructions (#29)
kairos-io/kairos-docs: content/en/docs/Installation/nvidia_agx_orin.md

---
title: "Nvidia AGX Orin"
linkTitle: "Nvidia AGX Orin"
weight: 4
date: 2022-11-13
description: >
  Install Kairos on Nvidia AGX Orin
---

This page describes how to install Kairos on [Nvidia AGX Orin](https://www.nvidia.com/en-us/autonomous-machines/embedded-systems/jetson-orin/) in the eMMC.

## Prerequisites

- [Nvidia AGX Orin](https://www.nvidia.com/en-us/autonomous-machines/embedded-systems/jetson-orin/)
- An USB type-C cable
- A Linux host used to flash the Nvidia AGX Orin board
- Jetson linux SDK [download](https://developer.nvidia.com/embedded/jetson-linux)

You can find debugging information here: https://developer.ridgerun.com/wiki/index.php/NVIDIA_Jetson_Orin/In_Board/Getting_in_Board/Serial_Console

## Flashing

> We are going to write the partitions in the eMMC. In order to do this we will use the Nvidia SDK configured with a custom partitioning layout and the partitions we are going to write in the eMMC.

Author: ```suggestion
We are going to write the partitions in the eMMC. In order to do this we will use the Nvidia SDK configured with a custom partitioning layout.
```

---
title: "Nvidia AGX Orin"
linkTitle: "Nvidia AGX Orin"
weight: 4
date: 2022-11-13
description: >
  Install Kairos on Nvidia AGX Orin
---

This page describes how to install Kairos on [Nvidia AGX Orin](https://www.nvidia.com/en-us/autonomous-machines/embedded-systems/jetson-orin/) in the eMMC.

## Prerequisites

- [Nvidia AGX Orin](https://www.nvidia.com/en-us/autonomous-machines/embedded-systems/jetson-orin/)
- An USB type-C cable
- A Linux host used to flash the Nvidia AGX Orin board
- Jetson linux SDK [download](https://developer.nvidia.com/embedded/jetson-linux)

You can find debugging informations here: https://developer.ridgerun.com/wiki/index.php/NVIDIA_Jetson_Orin/In_Board/Getting_in_Board/Serial_Console

## Flashing

> We are going to write the partitions in the eMMC. In order to do this we will use the Nvidia SDK configured with a custom partitioning layout and the partitions we are going to write in the eMMC.

Author: :+1: right, reworded it!

The partitions are:
- OEM for storing cloud config files (`/oem`)
- COS_STATE for storing the active/passive images to boot the system
- EFI for storing the efi shell and grub to boot the system
- RECOVERY - to store the recovery system

> - PERSISTENT - this is an optional partition to store the persistent data of the system. you can either write this in the eMMC or, for instance, to an external storage. It is enough to create a partition and label it as `COS_PERSISTENT`. There can be only one partition with such label. first match wins.

Reviewer: ```suggestion
- PERSISTENT - this is an optional partition to store the persistent data of the system. you can either write this in the eMMC or, for instance, to an external storage. It is enough to create a partition and label it as `COS_PERSISTENT`. There can be only one partition with such label, the first that matches wins.
```

### Prepare the SDK

The Jetson Linux SDK is used to perform the flashing process.

Download the Jetson Linux SDK:

```bash
wget https://developer.nvidia.com/downloads/embedded/l4t/r35_release_v3.1/release/jetson_linux_r35.3.1_aarch64.tbz2 -O tegra.bz2
tar xvf tegra.bz2

> 

Reviewer: ```suggestion
```

```

> We are going now to prepare the rootfs and the bootloader. The Jetson Linux SDK requires the rootfs to generate the `system.img` file and continue the flashing process however, we will not use the image generated by the SDK as we will use a different set of images (see below). Here we also disable extlinux as Kairos uses GRUB:

Reviewer: ```suggestion
Now, we are going to prepare the rootfs and the bootloader. The Jetson Linux SDK requires the rootfs to generate the `system.img` file and continue the flashing process however, we will not use the image generated by the SDK as we will use a different set of images (see below). Here we also disable extlinux as Kairos uses GRUB:
```

```bash
cd Linux_for_Tegra
# Drop extlinux
echo "" > ./bootloader/extlinux.conf
# This is needed so the SDK doesn't complain of missing files (not really used in the flash process)
IMAGE=quay.io/kairos/core-ubuntu-20-lts-arm-nvidia-jetson-agx-orin:latest
docker run -ti --rm -v $PWD/rootfs:/rootfs quay.io/luet/base util unpack "$IMAGE" /rootfs
# workaround needed (SDK writes to the symlink)
rm rootfs/boot/initrd

> ## Extlinux is required by the SDK - so we fake it in our root (it will not be there eventually)

Reviewer: ```suggestion
# Extlinux is required by the SDK - so we fake it in our root (it will not be there eventually)
```

mkdir -p rootfs/boot/extlinux/
echo "" > rootfs/boot/extlinux/extlinux.conf
```

### Prepare the images

> You can either download pre-built Kairos images, or build your own from a container image. You can find Kairos core ubuntu images based on Ubuntu `20.04` are here: https://quay.io/repository/kairos/core-ubuntu-20-lts-arm-nvidia-jetson-agx-orin.

Reviewer: ```suggestion
You can either download pre-built Kairos images, or build your own from a container image. You can find Kairos core ubuntu images based on Ubuntu `20.04` here: https://quay.io/repository/kairos/core-ubuntu-20-lts-arm-nvidia-jetson-agx-orin.
```

If you have instead the rootfs as a directory, you can create the required partitions with:
```bash
PATH=/rootfs/path
docker run --privileged \
        -e directory=$PATH \
        -e STATE_SIZE="6200" \
        -e RECOVERY_SIZE="4200" \
        -e DEFAULT_ACTIVE_SIZE="2000" \
        -v $PWD/bootloader:/bootloader --entrypoint /prepare_arm_images.sh -ti --rm quay.io/kairos/osbuilder-tools
```

{{% /tab %}}
{{< /tabpane >}}

After running any of the commands above, the generated images files required for flashing will be inside the `bootloader` directory (`bootloader/efi.img`, `bootloader/recovery_partition.img`, `bootloader/state_partition.img`, `bootloader/oem.img`, `bootloader/persistent.img` ).

{{% alert title="Note" %}}
The persistent image is optional, as you can store the system persistent data rather in an SD card or an NVME disk. The default `persistent.img` is of 2GB size. To create a persistent image manually of the size you prefer instead you can run:

```
# Create a 2GB filesystem for COS_PERSISTENT volume
truncate -s $((2048*1024*1024)) bootloader/persistent.img
mkfs.ext2 -L "COS_PERSISTENT" bootloader/persistent.img
```

> Note that the size of the partitions you modify should be dutely reported in the partition layout (see below).

Reviewer: ```suggestion
Note that the size of the partitions you modify should be duly reported in the partition layout (see below).
```
